File: typescript/src/io.ts
import { readFile, writeFile } from "node:fs/promises";
import { A3, A3ParseError } from "./a3";

/**
 * Read an A3 JSON file from disk and return a validated A3 instance.
 * @param path - Absolute or relative path to a `.a3.json` file.
 * @throws {@link A3ParseError} if the file cannot be read or is not valid JSON.
 * @throws {@link A3ValidationError} if the parsed content fails schema validation.
 */
export async function readJSON(path: string): Promise<A3> {
  let text: string;
  try {
    text = await readFile(path, "utf-8");
  } catch (e) {
    throw new A3ParseError(`Cannot read file: ${path}`, { cause: e });
  }
  return A3.fromJSONText(text);
}

/**
 * Write an A3 instance to a JSON file.
 * @param a3 - The A3 instance to serialize.
 * @param path - Absolute or relative path to write.
 * @param indent - Number of spaces for indentation (default 2). Pass 0 for compact output.
 * @throws {@link A3ParseError} if the file cannot be written.
 */
export async function writeJSON(a3: A3, path: string, indent = 2): Promise<void> {
  const text = a3.toJSONString(indent);
  try {
    await writeFile(path, text, "utf-8");
  } catch (e) {
    throw new A3ParseError(`Cannot write file: ${path}`, { cause: e });
  }
}

File: python/rtemis_a3/src/rtemis_a3/io.py
"""File I/O for A3 JSON files."""

from __future__ import annotations

from pathlib import Path

from ._models import A3
from .api import a3_from_json, a3_to_json
from .errors import A3ParseError


def read_a3json(path: str | Path) -> A3:
    """Read an A3 JSON file from disk.

    Parameters
    ----------
    path : str or Path
        Path to the JSON file.

    Returns
    -------
    A3
        Validated, immutable A3 object.

    Raises
    ------
    A3ParseError
        If file reading or JSON parsing fails.
    A3ValidationError
        If parsed data fails validation.
    """
    filepath = Path(path)
    try:
        text = filepath.read_text(encoding="utf-8")
    except OSError as exc:
        raise A3ParseError(f"cannot read file '{filepath}': {exc}") from exc
    return a3_from_json(text)


def write_a3json(a3: A3, path: str | Path, *, indent: int | None = None) -> None:
    """Write an A3 object to a JSON file.

    Parameters
    ----------
    a3 : A3
        A3 object to write.
    path : str or Path
        Output file path.
    indent : int, optional
        JSON indentation level. ``None`` for compact output.
    """
    filepath = Path(path)
    text = a3_to_json(a3, indent=indent)
    filepath.write_text(text + "\n", encoding="utf-8")
